Struqture doc refactor (#4)
kbarkhqs/struqture: documentation/src/physical_types/bosons/noisefree.md

## Operators and Hamiltonians

Complex objects are constructed from operator products are `BosonOperators` and `BosonHamiltonians`

> (for more information, [see also](../../container_types/operators_hamiltonians_and_systems.md)).

Reviewer: This sentence doesn't make sense

> Complex objects are constructed from operator products are `BosonOperators` and `BosonHamiltonians`
> (for more information, [see also](../../container_types/operators_hamiltonians_and_systems.md)).
> 
> These `BosonOperators` and `BosonHamiltonians` represent operators or Hamiltonians such as:

Reviewer: ```suggestion
`BosonOperators` and `BosonHamiltonians` represent operators or Hamiltonians such as:
```

\\[ \hat{O} = \sum_{j=0}^N \alpha_j \left( \prod_{k=0}^N f(j, k) \right) \left( \prod_{l=0}^N g(j, l) \right) \\]
with
\\[ f(j, k) = \begin{cases} c_k^{\dagger} \\\\ \mathbb{1} \end{cases} , \\]
\\[ g(j, l) = \begin{cases} c_l \\\\ \mathbb{1} \end{cases} , \\]
and 
\\(c^{\dagger}\\) the bosonic creation operator, \\(c\\) the bosonic annihilation operator 
\\[ \lbrack c_k^{\dagger}, c_j^{\dagger} \rbrack = 0, \\\\
    \lbrack c_k, c_j \rbrack = 0, \\\\
    \lbrack c_k^{\dagger}, c_j \rbrack = \delta_{k, j}. \\]

From a programming perspective the operators and Hamiltonians are HashMaps or Dictionaries with `BosonProducts` or `HermitianBosonProducts` (respectively) as keys and the coefficients \\(\alpha_j\\) as values. 

In `struqture` we distinguish between bosonic operators and Hamiltonians to avoid introducing unphysical behaviour by accident.

> While both are sums over normal ordered bosonic products (stored as HashMaps of products with a complex prefactor), Hamiltonians are guaranteed to be hermitian. In a bosonic Hamiltonian , this means that the sums of products are sums of hermitian bosonic products (we have not only the \\(c^{\dagger}c\\) terms but also their hermitian conjugate) and the on-diagonal terms are required to have real prefactors. 

Reviewer: ```suggestion
While both are sums over normal ordered bosonic products (stored as dictionaries of products with a complex prefactor), Hamiltonians are guaranteed to be hermitian. In a bosonic Hamiltonian , this means that the sums of products are sums of hermitian bosonic products (we have not only the \\(c^{\dagger}c\\) terms but also their hermitian conjugate) and the on-diagonal terms are required to have real prefactors. 
```

> In the `HermitianBosonProducts`, we only explicitly store one part of the hermitian bosonic product, and we have chosen to store the one which has the smallest index of the creators that is smaller than the smallest index of the annihilators.

Reviewer: Could you add an example here? (in scientific notation, not struqture notation)

Author: Added for Boson and Fermion

Reviewer: I don't see the example

### Example

> Here is an example of how to build a `BosonOperator` and a `BosonHamiltonian`:

Reviewer: This shows only how to build a bosonhamiltonian, could you first add an example for a bosonoperator?

```python
from struqture_py import bosons

# We start by initializing our BosonHamiltonian
hamiltonian = bosons.BosonHamiltonian()

# We create a BosonProduct or HermitianBosonProduct
bp = bosons.BosonProduct([0, 1], [0, 2])

> hbp = bosons.HermitianBosonProduct([0, 1], [0, 2])

Reviewer: Could you re-write these as strings?

Reviewer: (and then pass them directly in the set/add_... functions)

# Operators and Hamiltonians

`BosonOperators` and `BosonHamiltonians` represent operators or Hamiltonians such as:
\\[ \hat{O} = \sum_{j=0}^N \alpha_j \left( \prod_{k=0}^N f(j, k) \right) \left( \prod_{l=0}^N g(j, l) \right) \\]
with
\\[ f(j, k) = \begin{cases} b_k^{\dagger} \\\\ \mathbb{1} \end{cases} , \\]
\\[ g(j, l) = \begin{cases} b_l \\\\ \mathbb{1} \end{cases} , \\]
and 
\\(b^{\dagger}\\) the bosonic creation operator, \\(c\\) the bosonic annihilation operator 
\\[ \lbrack b_k^{\dagger}, b_j^{\dagger} \rbrack = 0, \\\\
    \lbrack b_k, b_j \rbrack = 0, \\\\
    \lbrack b_k^{\dagger}, b_j \rbrack = \delta_{k, j}. \\]

From a programming perspective the operators and Hamiltonians are HashMaps or Dictionaries with `BosonProducts` or `HermitianBosonProducts` (respectively) as keys and the coefficients \\(\alpha_j\\) as values. 

In `struqture` we distinguish between bosonic operators and Hamiltonians to avoid introducing unphysical behaviour by accident.

> While both are sums over normal ordered bosonic products (stored as dictionaries of products with a complex prefactor), Hamiltonians are guaranteed to be hermitian. In a bosonic Hamiltonian , this means that the sums of products are sums of hermitian bosonic products (we have not only the \\(b^{\dagger}b\\) terms but also their hermitian conjugate) and the on-diagonal terms are required to have real prefactors. 

Reviewer: ```suggestion
While both are sums over normal ordered bosonic products (stored as dictionaries of products with a complex prefactor), Hamiltonians are guaranteed to be hermitian. In a bosonic Hamiltonian, this means that the sums of products are sums of hermitian bosonic products (we have not only the \\(b^{\dagger}b\\) terms but also their hermitian conjugate) and the on-diagonal terms are required to have real prefactors. 
```

> In the `HermitianBosonProducts`, we only explicitly store one part of the hermitian bosonic product, and we have chosen to store the one which has the smallest index of the creators that is smaller than the smallest index of the annihilators. If the user choses \\(b_0^{\dagger}b_0\\) `HermitianBosonProduct([], [0])` will be created while the second part will be stored explicitly.

Reviewer: ```suggestion
In the `HermitianBosonProducts`, we only explicitly store one part of the hermitian bosonic product, and we have chosen to store the one which has the smallest index of the creators that is smaller than the smallest index of the annihilators. For instance, if the user would like to define a  \\(b_0^{\dagger} + b_0\\) term, they would create this object: `HermitianBosonProduct([], [0])`. The second part of the term is stored implicitly by the code.
```

In `struqture` we distinguish between bosonic operators and Hamiltonians to avoid introducing unphysical behaviour by accident.
While both are sums over normal ordered bosonic products (stored as HashMaps of products with a complex prefactor), Hamiltonians are guaranteed to be hermitian. In a bosonic Hamiltonian , this means that the sums of products are sums of hermitian bosonic products (we have not only the \\(c^{\dagger}c\\) terms but also their hermitian conjugate) and the on-diagonal terms are required to have real prefactors. 
In the `HermitianBosonProducts`, we only explicitly store one part of the hermitian bosonic product, and we have chosen to store the one which has the smallest index of the creators that is smaller than the smallest index of the annihilators.

## Example

Here is an example of how to build a `BosonOperator`:

```python
from struqture_py import bosons

# We start by initializing our BosonOperator
operator = bosons.BosonOperator()

# We create a BosonProduct or HermitianBosonProduct
bp = bosons.BosonProduct.from_string("c0c1a0a2")

> hbp = bosons.HermitianBosonProduct.from_string("c0c1a0a2")

Reviewer: ```suggestion
```

# We set the term and some value of our choosing

> operator.set(bp, 1.0 + 1.5j)

Reviewer: ```suggestion
operator.set("c0c1a0a2", 1.0 + 1.5j)
```

# We can use the `get` function to check what value/prefactor is stored for the BosonProduct

> assert operator.get(bp) == complex(1.0, 1.5)

Reviewer: ```suggestion
assert operator.get("c0c1a0a2") == complex(1.0, 1.5)
```

print(operator)

# Please note that the `set` function will set the value given, overwriting any previous value.
# Should you prefer to use and additive method, please use `add_operator_product`:

> operator.add_operator_product(bp, 1.0)

Reviewer: ```suggestion
operator.add_operator_product("c0c1a0a2", 1.0)
```
